### DISCUSSION TITLE
How to write livecheck for tencent-meeting?

### DISCUSSION THREAD

**Discussion by moonfruit** [Homebrew/discussions#1354] [Casks]
I am wondering how to write `livecheck` for cask [tencent-meeting](https://github.com/Homebrew/homebrew-cask/blob/master/Casks/tencent-meeting.rb).

There are some difficulties:

1. After digging the Tencent Meeting website, I found that to retrieve the latest version it needs to a HTTP POST request to specific URL. Like this:

   ```http
   POST https://meeting.tencent.com/wemeet-webapi/v2/config/query-download-info
   Content-Type: application/json
   
   [{"instance":"mac","type":"0300000000"}]
   ```

   And it returns:

   ```json
   {
     "code": 0,
     "message": "success",
     "nonce": "607cf3da39c67",
     "timestamp": 1618801626,
     "data": [
       {
         "md5": "b151cc342f235237a2702235d2220208",
         "url": "https://updatecdn.meeting.qq.com/b151cc342f235237a2702235d2220208/TencentMeeting_0300000000_2.9.4.401.publish.x86_64.dmg",
         "size": 95731756,
         "version": "2.9.4.401",
         "instance": "mac",
         "channel": "0300000000",
         "sub_date": "2021-04-01"
       }
     ]
   }
   ```

2. How to parse the JSON the HTTP request returns. Or use regex?
3. There are two versions for Tencent Meeting. One for x86, and another for arm64. Should I write for each of the versions? And the HTTP request I mentioned can also support retrieving the two versions in one request. Is there a way to use one `livecheck` for the two versions?

Any suggestion?

**Comment by carlocab** [comment]
CC @samford @nandahkrishna

**Reply by nandahkrishna** [reply]
This is unfortunately going to be similar to `tencent-docs`, as `livecheck` current doesn't have a strategy that does POST requests. From that Cask:

```
livecheck do
    # curl \
    #   -H 'content-type: application/json' \
    #   --data-binary \
    #   '{"pull_item":{"app_id":"936c3190-4a5a-4608-ac9e-d37fe33f40be","group":"DOWNLOAD.release-note-mac"}}' \
    #   https://docs.qq.com/config.v2.ConfigService/PullConfigReq
    url "https://docs.qq.com/home/download"
    strategy :page_match
    regex(/Mac\s*版本号\s*(\d+(?:\.\d+)*)/i)
    skip "only works with POST request"
  end
```

Sam would probably have more to say about whether this feature would be added (or exists in some form and I'm unaware).

**Reply by samford** [reply]
As Nanda mentioned, livecheck only supports GET requests at the moment. I think the way to support POST requests would be to allow for defining optional configuration information in the `livecheck` block and passing this to `Strategy#page_content` (or `Strategy#page_headers`) to modify the behavior of the underlying `curl` call. Approaching it this way would allow us to control how content is fetched regardless of which strategy is used.

I have some in-progress work stashed locally that adds support for passing options to the aforementioned `Strategy` methods, each strategy's `#find_versions` method, etc. Basically, it's a general approach that will allow us to provide optional configuration information to various parts of livecheck as needed. Support for POST requests can be built on this but I have some other work that I need to wrap up before I create a PR for `options` support in `livecheck` blocks.

---

Similarly, I've been thinking about adding general-purpose strategies for certain data formats like `JSON`, `XML`, etc. This would allow us to work with the parsed data in a `strategy` block (though some existing strategies could also be modified to benefit) instead of being restricted to a regex. It would potentially be more precise than using a regex but it also may be a bit more cumbersome, so it's something that I want to experiment with and explore first.

For the time being, we parse JSON data using a regex (e.g., `/"version":\s*?"v?(\d+(?:\.\d+)+)"/i`) and this tends to be fine when we're only interested in one field.  In this case, you would have to use a `strategy` block with multiple regexes to assemble the `version` and `md5` JSON fields into one version string (this isn't uncommon for casks).

---

I'm not a cask expert but I believe it may be possible to have separate `livecheck` blocks in the `if`/`else` branches. However, this would only return the version information for the architecture that you're running `brew livecheck` on. If you need the version information for multiple architectures to update the cask, then we run into another limitation in livecheck. Currently livecheck is built around the idea of only returning one version and it would take some work to change that.

Around 55 casks currently contain `if Hardware::CPU.intel?` but I imagine that number may grow a bit over time. It's not a huge number but it seems like we'll need to determine how to handle this situation sooner or later. I'm busy with some other areas of livecheck at the moment but I'll try to remember to open an issue in Homebrew/brew when I get a chance and we can collectively discuss approaches.

---

All that said, I agree with Nanda that the best solution for now is to simply `skip` this cask like `tencent-docs`. If/when we've cleared the technical hurdles in the future, we can come back to these casks and create a working `livecheck` block.

**Reply by moonfruit** [reply]
Thanks for the reply, I'm looking forward to the improvement of the `livecheck`, and willing to do something about this.
